### DISCUSSION TITLE
Custom reticle not updating position at ray's end

### DISCUSSION THREAD

**Discussion by nRammm** [microsoft/MixedRealityToolkit-Unity#11751] [Q&A] [answered]
I've spawned a custom reticle using BaseReticleVisual's `AttachCustomReticle(GameObject reticleInstance)` . Able to see the custom reticle instead of the default one, but the position of the reticle is not updating to the ray hit/end point rather it is at the ray origin point always. How can I make it to update properly? 

 I'm basically trying to achieve this in "all valid surfaces" instead of "interactables only" and I've configured the visibility settings of MRTKRayReticleVisual in this way.

![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/93316456/39177d4d-dd45-4f9b-8a45-0b5839da9679)

Also sometimes the reticle is coming up only for left hand or for right hand. I couldn't see them in both the hands.
![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/93316456/ebf7d475-0310-4e49-9b42-c506ef57dead)

Is there anything I'm doing wrong ?

Thanks.

**Comment by nRammm** [comment]
@AMollis , can I get any suggestion on the above query?
Thanks.

**Reply by AMollis** [reply]
XRI interactions only target interactables.  You need to add an interactable component to the colliders you want to interact with.  XRI is optimized to only send interaction events to game object's whose colliders that have been registered with the interaction manager (see [XRInteractionManager::RegisterInteractable](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRInteractionManager.html#UnityEngine_XR_Interaction_Toolkit_XRInteractionManager_RegisterInteractable_UnityEngine_XR_Interaction_Toolkit_IXRInteractable_)).  An XR interactable component handles this registration for you.

What I suspect is happening is that XRRayInteractor is colliding with your object (see XRRayInteractor::CheckCollidersBetweenPoints), but then [XRInteractionManager::TryGetInteractableForCollider](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@1.0/api/UnityEngine.XR.Interaction.Toolkit.XRInteractionManager.html#UnityEngine_XR_Interaction_Toolkit_XRInteractionManager_TryGetInteractableForCollider_Collider_) filters it out, since the collider doesn't belong to an interactable.

You really need to add an interactable component to the things you want to be interacted with. Perhaps reading through the XRI architecture will assist here:

https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/manual/architecture.html

**Reply by nRammm** [reply]
Thanks for the reply. Will look into the references you provided.

 Actually I've added interactable component to those elements in the scene . If you could see the image above , those blue coloured buttons are MRTK's PressableButtons. But still on hovering them I could see that the custom reticle is located at ray's origin only(its not at the colliding point). That's what I'm wondering why its not updating for the Pressable button as such.
 
 One observation related to this is ,  the custom reticle(see the img below "RingVisual Variant is the custom reticle I used") is placed as a child of FarRay transform , but the position and rotation is applied to the RayReticle transform(ReticleRoot of MRTKRayReticleVisual) right? Shouldn't the custom reticle be placed as child of this root to update the position and rotation properly? 
![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/93316456/81202358-453c-4a36-b62e-f36059c9f9e4)

**Reply by AMollis** [reply]
I see.   At the moment it is not attached to the transform.  

To your custom reticle object add a component that implements  Microsoft.MixedReality.Toolkit.IVariableReticle, and you will receive the reticle position that you can copy to your transform.

For example,

```
    public class CustomReticle : MonoBehaviour, IVariableReticle
    {
        public void UpdateVisuals(VariableReticleUpdateArgs args)
        {
            transform.SetPositionAndRotation(args.ReticlePosition, Quaternion.LookRotation(args.ReticleNormal, Vector3.up));
        }
    }
```

The `UpdateVisuals` function is called every frame.

Arguably this can be made a bit easier, and we can look at ways to simplify custom reticles.

**Reply by AMollis** [reply]
@nRammm, you can try the workaround I provided above...

Or wait for this PR to complete

- https://github.com/microsoft/MixedRealityToolkit-Unity/pull/11758

Hopefully the PR makes things easier for you.

**Reply by nRammm** [reply]
Sounds great! Thanks for the help @AMollis .

I hope this fixes the reticle position issue. But on the other hand , I'm trying to attach custom reticle via script(see the snippet below) , I'm populating interactorsCache with `interactionManager.GetRegisteredInteractors(interactorsCache)` .But on doing so, I couldn't see the custom reticle(or even the default one) appearing on any one of the hands(sometimes its not seen in left and sometimes not in right , I'm not sure why its happening eventhough it hits `AttachCustomReticle` for both the left and right ray interactor) . Do you think this approach serves my expectation to bring up custom reticle for both the hands?

![image](https://github.com/microsoft/MixedRealityToolkit-Unity/assets/93316456/54ac3a5e-648e-4cc1-b7ba-b1763870720a)

**Reply by nRammm** [reply]
Hi @AMollis , any suggestion on the above way of populating custom reticle?

**Reply by AMollis** [reply, marked as answer]
You're adding the same game object to all interactors.   Use `customCursor` as a prefab and create new instances of it.  Also, you can probably change the code to not rely on MRTK3.

```
if (customCursor != null)
{
  foreach (var interactor in interactorCache)
  {
      var reticleProvider = interactor.transform.GetComponent<IXRCustomReticleProvider>();
      if (reticleProvider != null}
      {
         var instance = Instantiate(customCursor);
         reticleProvider.AttachCustomReticle(instance);
      }
  }
}
```

**Comment by cheitovilla** [comment]
Honestly, it's amazing
